### PR TITLE
traffic shaping: take configuration via a runtimeConfig

### PR DESCRIPTION
**Pull request by m1093782566**
To make it easier to supply bandwidth in a dynamic manner, the traffic shaping plugin should support a capability arg.

Fixes: #133

### CODE REVIEW

**Comment by m1093782566** [comment]
@squeed @caseydavenport

**Comment by m1093782566** [comment]
```
• Failure [0.796 seconds]
Basic PTP using cnitool when the bandwidth plugin is chained with a plugin that returns multiple adapters [Measurement] limits traffic only on the restricted bandwith veth device 
/home/travis/gopath/src/github.com/containernetworking/plugins/gopath/src/github.com/containernetworking/plugins/integration/integration_linux_test.go:151
  Expected
      <time.Duration>: 26698362
  to be >
      <time.Duration>: 1023954067
  /home/travis/gopath/src/github.com/containernetworking/plugins/gopath/src/github.com/containernetworking/plugins/integration/integration_linux_test.go:188
```

The failed test seems flaky, so how to re-trigger the test?

**Comment by dcbw** [comment]
@m1093782566 restarted...

**Comment by dcbw** [comment]
@m1093782566 I think it's useful to have both runtime config and static config; also allows limiting if the runtime doesn't yet support the dynamic modes.  Other than that, looks OK...

**Comment by Lion-Wei** [comment]
@dcbw By static config you mean config like this?
```json
{
            "type": "bandwidth",
            "ingressRate": 800,
            "ingressBurst": 200,
            "egressRate": 800,
            "egressBurst": 200
}
```

**Comment by m1093782566** [comment]
@dcbw CI is green now.

**Comment by dcbw** [comment]
@Lion-Wei yeah, like that.

**Comment by squeed** [comment]
I agree with @dcbw - the plugin should take both static and runtime configuration

**Comment by Lion-Wei** [comment]
@dcbw @squeed , make sense. We can do that. : )

File: plugins/meta/bandwidth/main.go [hunk 1]
```diff
@@ -135,9 +154,20 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return err
 	}
 
-	//no traffic shaping was requested, so just no-op and quit
+	var bandwidth *BandWidthEntry
 	if conf.IngressRate == 0 && conf.IngressBurst == 0 && conf.EgressRate == 0 && conf.EgressBurst == 0 {
-		return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+		bandwidth = conf.RuntimeConfig.BandWidth
+		//no traffic shaping was requested, so just no-op and quit
+		if bandwidth == nil || (bandwidth.IngressRate == 0 && bandwidth.IngressBurst == 0 && bandwidth.EgressRate == 0 && bandwidth.EgressBurst == 0) {
```
Comment (squeed) [inline plugins/meta/bandwidth/main.go]: I suggest you add an `IsZero` method on the BandwidthEntry type - makes these comparisons simpler.

File: plugins/meta/bandwidth/main.go [hunk 2]
```diff
@@ -17,32 +17,42 @@ package main
...
+
+	RuntimeConfig struct {
+		BandWidth *BandWidthEntry `json:"bandWidth,omitempty"`
+	} `json:"runtimeConfig,omitempty"`
+
+	// RuntimeConfig *struct{} `json:"runtimeConfig"`
+
+	RawPrevResult *map[string]interface{} `json:"prevResult"`
+	PrevResult    *current.Result         `json:"-"`
+	BandWidthEntry
```
Comment (squeed) [inline plugins/meta/bandwidth/main.go]: The nested struct can be a pointer, and the json deserializer will handle it correctly - makes it easy to track empty values.

Comment (squeed) [reply plugins/meta/bandwidth/main.go]: We do something similar with the host-local IPAM, if you want to see it in action.

File: plugins/meta/bandwidth/main.go (hunk 1 above)
Comment (squeed) [reply plugins/meta/bandwidth/main.go]: Also, if you make the nested BandwidthEntry struct in config a pointer, this can be simpler.

**Comment by squeed** [comment]
Minor request: The "w" should be lowercase everywhere, including config.

**Comment by squeed** [comment]
A few minor things, but otherwise looks good! When you're all done, can you squash to a single commit (and force-push to your branch)?

**Comment by m1093782566** [comment]
@squeed I squashed the commits and it still have two commits: one is mine and the other one is my friend @Lion-Wei 's as I paired with @Lion-Wei on this work :)

File: plugins/meta/bandwidth/main.go (hunk 2 above)
Comment (Lion-Wei) [reply plugins/meta/bandwidth/main.go]: Good Point, thanks, Now wit's can be much simpler, PTAL. : )

**Comment by m1093782566** [comment]
@squeed @dcbw 

All comments are addressed now. PTAL. BTW, the work of having Kubernetes CNI network driver support traffic shaping still relies on this PR.

**Comment by squeed** [comment]
Looks good to me. Is there any reason you changed the bandwidth values in the integration test? Nice job, btw.

**Comment by m1093782566** [comment]
@squeed 

Oops... It is changed back now. Please take a look again when you have a chance, thanks!

**Comment by squeed** [comment]
LGTM! Thanks! Now just need a second reviewer.

**Comment by m1093782566** [comment]
Thanks @squeed 

@dcbw would you please take a look?

**Comment by rosenhouse** [comment]
back ref: https://github.com/kubernetes/kubernetes/pull/63194

**Pull request merged**